### PR TITLE
EN-13432: fixes for nodes coordinator with chain parameters

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
## Reasoning behind the pull request
- previously, current chain parameters were used, leading to unexpected edge cases at epoch-change events
  
## Proposed changes
- use the function that computes the chain parameters based on the provided epoch

## Testing procedure
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/ElrondNetwork/elrond-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: epochStart/metachain/rewards_test.go [hunk 1]
```diff
@@ -668,15 +668,15 @@ func TestRewardsCreator_ProtocolRewardsForValidatorFromMultipleShards(t *testing
...
+	rwdc.fillBaseRewardsPerBlockPerNode(mb.EpochStart.Economics.RewardsPerBlock, 0)
 	rwdInfoData := rwdc.computeValidatorInfoPerRewardAddress(valInfo, &rewardTx.RewardTx{}, 0)
 	assert.Equal(t, 1, len(rwdInfoData))
 	rwdInfo := rwdInfoData[pubkey]
 	assert.Equal(t, rwdInfo.address, pubkey)
 
 	assert.Equal(t, rwdInfo.accumulatedFees.Cmp(big.NewInt(200)), 0)
-	protocolRewards := uint64(valInfo[0][0].NumSelectedInSuccessBlocks) * (mb.EpochStart.Economics.RewardsPerBlock.Uint64() / uint64(args.NodesConfigProvider.ConsensusGroupSize(0)))
-	protocolRewards += uint64(valInfo[core.MetachainShardId][0].NumSelectedInSuccessBlocks) * (mb.EpochStart.Economics.RewardsPerBlock.Uint64() / uint64(args.NodesConfigProvider.ConsensusGroupSize(core.MetachainShardId)))
+	protocolRewards := uint64(valInfo[0][0].NumSelectedInSuccessBlocks) * (mb.EpochStart.Economics.RewardsPerBlock.Uint64() / uint64(args.NodesConfigProvider.ConsensusGroupSize(0, 0)))
```
Comment (iulianpascalau) [inline epochStart/metachain/rewards_test.go:678]: this is a long line along with the following. Could you extract a var for the consensusGroupSize?

File: factory/consensus/consensusComponents.go [hunk 2]
```diff
@@ -238,7 +238,9 @@ func (ccf *consensusComponentsFactory) Create() (*consensusComponents, error) {
 	// apply consensus group size on the input antiflooder just before consensus creation topic
 	ccf.networkComponents.InputAntiFloodHandler().ApplyConsensusSize(
 		ccf.processComponents.NodesCoordinator().ConsensusGroupSize(
```
Comment (iulianpascalau) [inline factory/consensus/consensusComponents.go:240]: extract in a var the consensus group size?

File: sharding/chainParametersHolder.go [hunk 3]
```diff
@@ -48,9 +49,23 @@ func NewChainParametersHolder(args ArgsChainParametersHolder) (*chainParametersH
...
 	args.EpochStartEventNotifier.RegisterHandler(paramsHolder)
 
+	logInitialConfiguration(args.ChainParameters)
+
 	return paramsHolder, nil
 }
 
+func logInitialConfiguration(chainParameters []config.ChainParametersByEpochConfig) {
+	logMessage := "initialized chainParametersHolder with the values:\n"
+	logLines := make([]string, 0)
```
Comment (iulianpascalau) [inline sharding/chainParametersHolder.go:59]:
```GO
logLines := make([]string, 0, len(chainParameters))
```

File: sharding/chainParametersHolder.go [hunk 4]
```diff
@@ -48,9 +49,23 @@ func NewChainParametersHolder(args ArgsChainParametersHolder) (*chainParametersH
...
+	for _, params := range chainParameters {
+		logLines = append(logLines, fmt.Sprintf("\tenable epoch=%d, round duration=%d, hysteresis=%.2f, shard consensus group size=%d, shard min nodes=%d, meta consensus group size=%d, meta min nodes=%d, adaptivity=%v",
+			params.EnableEpoch, params.RoundDuration, params.Hysteresis, params.ShardConsensusGroupSize, params.ShardMinNumNodes, params.MetachainConsensusGroupSize, params.MetachainMinNumNodes, params.Adaptivity))
+	}
+
+	logMessage += strings.Join(logLines, "\n")
+	log.Debug(logMessage)
```
Comment (iulianpascalau) [inline sharding/chainParametersHolder.go:66]: :+1:

File: epochStart/interface.go [hunk 5]
```diff
@@ -129,7 +129,7 @@ type StartOfEpochMetaSyncer interface {
 
 // NodesConfigProvider will provide the necessary information for start in epoch economics block creation
 type NodesConfigProvider interface {
-	ConsensusGroupSize(shardID uint32) int
+	ConsensusGroupSize(shardID uint32, epoch uint32) int
```
Comment (iulianpascalau) [inline epochStart/interface.go:132]: I believe this is a bit dangerous as I can call the function as
```GO
NodesConfigProvider.ConsensusGroupSize(epoch, shardID)
```
and the compiler will be just happy. Other than using a special argument DTO struct which will complicate things, the only option is to have the epoch as something else than a uint32. But that will need a lot of useless casts.

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go [hunk 6]
```diff
@@ -611,7 +614,13 @@ func (ihnc *indexHashedNodesCoordinator) EpochStartPrepare(metaHdr data.HeaderHa
 	unStakeLeavingList := ihnc.createSortedListFromMap(newNodesConfig.leavingMap)
 	additionalLeavingList := ihnc.createSortedListFromMap(additionalLeavingMap)
 
+	chainParamsForEpoch, err := ihnc.chainParametersHandler.ChainParametersForEpoch(newEpoch)
+	if err != nil {
+		log.Warn("indexHashedNodesCoordinator.EpochStartPrepare: could not compute chain params for epoch", "epoch", newEpoch, "error", err)
```
Comment (iulianpascalau) [inline sharding/nodesCoordinator/indexHashedNodesCoordinator.go:619]: maybe add the message here, on L1057 and in sharding/nodesCoordinator/indexHashedNodesCoordinatorLite.go : "... and will use the current chain parameters" as that is the fallback.

File: sharding/nodesCoordinator/indexHashedNodesCoordinator_test.go [hunk 7]
```diff
@@ -1568,6 +1516,14 @@ func TestIndexHashedNodesCoordinator_EpochStart_EligibleSortedAscendingByIndex(t
 					MetachainConsensusGroupSize: 1,
 				}
 			},
+			ChainParametersForEpochCalled: func(_ uint32) (config.ChainParametersByEpochConfig, error) {
```
Comment (iulianpascalau) [inline sharding/nodesCoordinator/indexHashedNodesCoordinator_test.go:1519]: maybe use an assert here to check the epoch is the same provided on L1848/L1849?

File: epochStart/metachain/rewards_test.go (hunk 1 above)
Comment (bogdan-rosianu) [reply epochStart/metachain/rewards_test.go:678]: extracted in more variables

File: sharding/chainParametersHolder.go (hunk 3 above)
Comment (bogdan-rosianu) [reply sharding/chainParametersHolder.go:59]: done

File: epochStart/interface.go (hunk 5 above)
Comment (bogdan-rosianu) [reply epochStart/interface.go:132]: renamed to `ConsensusGroupSizeForShardAndEpoch`

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go (hunk 6 above)
Comment (bogdan-rosianu) [reply sharding/nodesCoordinator/indexHashedNodesCoordinator.go:619]: added

File: sharding/nodesCoordinator/indexHashedNodesCoordinator_test.go (hunk 7 above)
Comment (bogdan-rosianu) [reply sharding/nodesCoordinator/indexHashedNodesCoordinator_test.go:1520]: added

File: sharding/nodesCoordinator/indexHashedNodesCoordinator_test.go [hunk 8]
```diff
@@ -1828,16 +1828,26 @@ func TestIndexHashedNodesCoordinator_GetConsensusWhitelistedNodesAfterRevertToEp
 func TestIndexHashedNodesCoordinator_ConsensusGroupSize(t *testing.T) {
 	t.Parallel()
 
+	testEpoch := uint32(37)
```
Comment (iulianpascalau) [inline sharding/nodesCoordinator/indexHashedNodesCoordinator_test.go:1831]: :+1:

**Review by iulianpascalau — DISMISSED** [review]

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go [hunk 9]
```diff
@@ -1037,11 +1047,21 @@ func (ihnc *indexHashedNodesCoordinator) computeShardForSelfPublicKey(nodesConfi
...
 
-// ConsensusGroupSize returns the consensus group size for a specific shard
-func (ihnc *indexHashedNodesCoordinator) ConsensusGroupSize(
+// ConsensusGroupSizeForShardAndEpoch returns the consensus group size for a specific shard in a given epoch
+func (ihnc *indexHashedNodesCoordinator) ConsensusGroupSizeForShardAndEpoch(
 	shardID uint32,
+	epoch uint32,
 ) int {
-	currentChainParameters := ihnc.chainParametersHandler.CurrentChainParameters()
+	ihnc.mutNodesConfig.RLock()
```
Comment (AdoAdoAdo) [inline sharding/nodesCoordinator/indexHashedNodesCoordinator.go:1055]: why is this locking required?
If it is only for the log ihnc.currentEpoch, I would change the log and remove the current epoch and remove the locking as well. I think more relevant for the warn would be to print the epoch for which the consensusGroupSize failed on.

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: epochStart/metachain/baseRewards.go [hunk 10]
```diff
@@ -430,15 +430,17 @@ func (brc *baseRewardsCreator) finalizeMiniBlocks(miniBlocks block.MiniBlockSlic
 	return finalMiniBlocks
 }
 
-func (brc *baseRewardsCreator) fillBaseRewardsPerBlockPerNode(baseRewardsPerNode *big.Int) {
+func (brc *baseRewardsCreator) fillBaseRewardsPerBlockPerNode(baseRewardsPerNode *big.Int, epoch uint32) {
+	// use previous epoch for fetching the consensus group size, since the epoch start metablock already contains the new epoch
+	epochForConsensusSize := epoch - 1
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply epochStart/metachain/baseRewards.go:435]: right. I've treated the epoch 0 scenario

File: sharding/nodesCoordinator/indexHashedNodesCoordinator.go (hunk 9 above)
Comment (bogdan-rosianu) [reply sharding/nodesCoordinator/indexHashedNodesCoordinator.go:1055]: right, good catch. fixed

**Review by AdoAdoAdo — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**